### DISCUSSION TITLE
nanoPi r6c crashes when starting containers (docker compose up -d)

### DISCUSSION THREAD

**Discussion by plix1014** [MichaIng/DietPi#7162] [Q&A]
I have a nanoPi R6C running with dietpi v9.6.1 (debian bookworm) with several docker containers with no issues. In December the nanopi started to crash during container startup. I could not figure out, what was the reason. After several retries, I could get my services up and running again. In fact, I replaced it for two weeks with the old raspberryi Pi4 pihole (no docker). When I switched back, It worked without any issue, although I did not change anything (shutdown unplug, replace with pi4, two weeks later unplug pi4 and connect nanopi and boot).
Since then I had zero issues. Now it starts again. I have no glue what is the reason. I just get disconnected when starting the containers.
Also on the console I do not see an obvious error.

see 01_docker_pihole_start.txt: docker compose 
see 02_nanopi_r6c_crash_reboot.txt: debug console output.

Any idea what could trigger the crash? 

I also have three raspberry pi (4,5) with docker containers (debian bookworm). Non of them crashed once when starting a container.

[02_nanopi_r6c_crash_reboot.txt](https://github.com/user-attachments/files/16254170/02_nanopi_r6c_crash_reboot.txt)
[01_docker_pihole_start.txt](https://github.com/user-attachments/files/16254172/01_docker_pihole_start.txt)

**Comment by MichaIng** [comment]
```
client_loop: send disconnect: Broken pipe
```
This error message is no crash and has nothing to do directly with Docker/the container: It is output from your SSH client, that it got connected from the OpenSSH server. You also see that you land on the client side console after the error. Check your OpenSSH config. Here is some info about the two relevant settings to check: https://www.tecmint.com/client_loop-send-disconnect-broken-pipe/
Probably the container startup takes very long?

**Reply by plix1014** [reply]
Well, if I'm logged in from remote, the last messages is the broken pipe, which indices that I lost the connection.
After relogin, I see, that the server rebooted. The container than cannot be started, because some "docker-proxy ..." processes are already running. So I need to kill them (systemctl restart docker). When I try to start the container, I get disconnected again.
The startup of the container, if it works, is quit quick, so no long startup time.

**Reply by MichaIng** [reply]
Sorry for the late reply. Does the issue persist?

Hmm, if the some docker-proxy service runs already, were the containers maybe started automatically at boot, together with the `docker.service`?

Maybe enabling persistent system logs can help to check what went wrong:
```sh
dietpi-software uninstall 103 # this uninstalls RAM logging
mkdir /var/log/journal # systemd logs to disk automatically if this dir exists
reboot
```

**Reply by plix1014** [reply]
I replaced the device (with a regular Pi5) as I needed the dns.
Yes, the docker service tried to start the containers which failed. I cleanup docker processes (docker compose down) and killed any other processes, restarted the docker.service and tried to startup pihole, which immediatly lead to a new restart. I did this many times.

This issue not only appeared with pihole but also other containers (e.g. vaultwarden, mediawiki). So this is not linked to a specific container.
I even removed all docker images (docker system prune; docker image rm xxxxx) and pulled the containers but I had no luck. I had this issue already a few months earlier and after a lot of retries the startup succeeded. I didn't know why it worked but hoped, that the issue will never reappear.
I was wrong, and this time, I couldn't start the container after several hours. So I decided to switch the hardware.

The interesting thing is, after some weeks where the nano pi was offline(completely disconnected from power and LAN) I connected the pi again (only changed the IP) and I could start the containers without issue. Since it is running in idle mode (no client is using it) this issue never appeared again.

The RAMlog was already uninstalled.
# dietpi-software uninstall 103
[  OK  ] DietPi-Software | Initialised database
[  OK  ] DietPi-Software | Reading database

 DietPi-Software
─────────────────────────────────────────────────────
 Mode: Automated uninstall

[ INFO ] DietPi-Software | 103: DietPi-RAMlog is not currently installed
[  OK  ] DietPi-Software | No changes applied for: DietPi-RAMlog

**Reply by MichaIng** [reply]
In the meantime, there have been a lot of kernel updates as well. In between, the kernel had an issue, limiting processes to 400 something, which has been solved just yesterday. Maybe this caused issues in your case as well.

In any case, if a crash happens, apart of disabling RAM logging, enabling persistent system/journal logs is essential, to have a chance to see what was going wrong. The systemd journal (=system log) by default goes to RAM/tmpfs (`/run/systemd/journal`), but it switches to log to `/var/log/journal`, if that directory exists.

And just another approach: I know some prefer to have everything in containers, but all software titles you mention can be easily installed bare metal via `dietpi-software`, which does not have such kind of intransparent container engine related startup issues. Containers have their use cases, especially if some software requires a complex environment and dependencies, which are difficult to achieve bare metal on a particular distro with fixed Python/PHP/whatever versions. But they imply a lot of complexity, and hence error sources, which are often not understood by users. DietPi/`dietpi-software` explicitly aims to combine the benefits of both, solving complex or specific dependency/setup needs for famous software titles with minimal overhead, no container engine in between. You might have particular reasons to use Docker, but otherwise I can only recommend to give it a try, and compare the result/usability and maintenance efforts in the mid/long-term.

**Reply by plix1014** [reply]
Thanks for the info. The main reason, why I use containers is easier migration from os version (or hw revision) to a new one.
Although I use ansible for configuration it is a lot of work to transfer from the old to the new system. With containers this is much less work.
